File: dotnet/tests/Snowflake.Data.Tests/IntTest.cs
using Snowflake.Data.Tests.Compatibility;

namespace Snowflake.Data.Tests;

[Trait("Category", "E2E")]
public class IntTest : IClassFixture<ITFixture>
{
    private static readonly string[] IntTypeSynonyms =
        ["INT", "INTEGER", "BIGINT", "SMALLINT", "TINYINT", "BYTEINT"];

    protected readonly ITestOutputHelper Output;
    protected readonly ITFixture Fixture;

    public IntTest(ITFixture fixture, ITestOutputHelper output)
    {
        Fixture = fixture;
        Output = output;
    }

    // Scenario: should cast integer values to appropriate type for int and synonyms
    [SnowflakeTheory]
    [InlineData("INT")]
    [InlineData("INTEGER")]
    [InlineData("BIGINT")]
    [InlineData("SMALLINT")]
    [InlineData("TINYINT")]
    [InlineData("BYTEINT")]
    public void ShouldCastIntegerValuesToAppropriateTypeForIntAndSynonyms(string intType)
    {
        // Given Snowflake client is logged in
        using var connection = Fixture.Factory.Create(Output);
        connection.Open();

        // When Query "SELECT 0::<type>, 1000000::<type>, 9223372036854775807::<type>" is executed
        using var cmd = connection.CreateCommand();
        cmd.CommandText = $"SELECT 0::{intType}, 1000000::{intType}, 9223372036854775807::{intType}";
        using var reader = cmd.ExecuteReader();

        // Then All values should be returned as appropriate type with no precision loss
        Assert.True(reader.Read(), "Expected one row");
        Assert.Equal(0L, reader.GetInt64(0));
        Assert.Equal(1000000L, reader.GetInt64(1));
        Assert.Equal(9223372036854775807L, reader.GetInt64(2));
        Assert.False(reader.Read(), "Expected exactly one row");
    }

    // Scenario Outline: should select integer <values> for int and synonyms
    [SnowflakeTheory]
    [MemberData(nameof(BoundaryValuesData))]
    public void ShouldSelectIntegerValuesForIntAndSynonyms(string intType, string values, long[] expectedValues)
    {
        _ = values; // used for test display name only

        // Given Snowflake client is logged in
        using var connection = Fixture.Factory.Create(Output);
        connection.Open();

        // When Query "SELECT <query_values>" is executed
        using var cmd = connection.CreateCommand();
        var castExpressions = expectedValues.Select(v => $"{v}::{intType}");
        cmd.CommandText = $"SELECT {string.Join(", ", castExpressions)}";
        using var reader = cmd.ExecuteReader();

        // Then Result should contain integers <expected_values>
        Assert.True(reader.Read(), "Expected one row");
        for (var i = 0; i < expectedValues.Length; i++)
        {
            Assert.Equal(expectedValues[i], reader.GetInt64(i));
        }
        Assert.False(reader.Read(), "Expected exactly one row");
    }

    public static IEnumerable<object[]> BoundaryValuesData()
    {
        foreach (var intType in IntTypeSynonyms)
        {
            yield return [intType, "zero", new long[] { 0 }];
            yield return [intType, "tinyint", new long[] { -128, 127, 255 }];
            yield return [intType, "smallint", new long[] { -32768, 32767, 65535 }];
            yield return [intType, "int", new long[] { -2147483648, 2147483647, 4294967295 }];
            yield return [intType, "bigint", new long[] { -9223372036854775808, 9223372036854775807 }];
        }
    }

    // Scenario: should handle NULL values for int and synonyms
    [SnowflakeTheory]
    [InlineData("INT")]
    [InlineData("INTEGER")]
    [InlineData("BIGINT")]
    [InlineData("SMALLINT")]
    [InlineData("TINYINT")]
    [InlineData("BYTEINT")]
    public void ShouldHandleNullValuesForIntAndSynonyms(string intType)
    {
        // Given Snowflake client is logged in
        using var connection = Fixture.Factory.Create(Output);
        connection.Open();

        // When Query "SELECT NULL::<type>, 42::<type>, NULL::<type>" is executed
        using var cmd = connection.CreateCommand();
        cmd.CommandText = $"SELECT NULL::{intType}, 42::{intType}, NULL::{intType}";
        using var reader = cmd.ExecuteReader();

        // Then Result should contain [NULL, 42, NULL]
        Assert.True(reader.Read(), "Expected one row");
        Assert.True(reader.IsDBNull(0), "Expected NULL for column 0");
        Assert.Equal(42L, reader.GetInt64(1));
        Assert.True(reader.IsDBNull(2), "Expected NULL for column 2");
        Assert.False(reader.Read(), "Expected exactly one row");
    }

    // Scenario: should download large result set with multiple chunks for int and synonyms
    [SnowflakeTheory]
    [InlineData("INT")]
    [InlineData("INTEGER")]
    [InlineData("BIGINT")]
    [InlineData("SMALLINT")]
    [InlineData("TINYINT")]
    [InlineData("BYTEINT")]
    public void ShouldDownloadLargeResultSetWithMultipleChunksForIntAndSynonyms(string intType)
    {
        const int rowCount = 50000;

        // Given Snowflake client is logged in
        using var connection = Fixture.Factory.Create(Output);
        connection.Open();

        // When Query "SELECT seq8()::<type> as id FROM TABLE(GENERATOR(ROWCOUNT => 50000)) v ORDER BY id" is executed
        using var cmd = connection.CreateCommand();
        cmd.CommandText = $"SELECT (ROW_NUMBER() OVER (ORDER BY seq8()) - 1)::{intType} AS id FROM TABLE(GENERATOR(ROWCOUNT => {rowCount})) v ORDER BY 1";
        using var reader = cmd.ExecuteReader();

        // Then Result should contain 50000 sequentially numbered rows from 0 to 49999
        var count = 0;
        while (reader.Read())
        {
            Assert.Equal((long)count, reader.GetInt64(0));
            count++;
        }
        Assert.Equal(rowCount, count);
    }

    // Scenario Outline: should select <values> from table for int and synonyms
    [SnowflakeTheory]
    [MemberData(nameof(TableValuesData))]
    public void ShouldSelectValuesFromTableForIntAndSynonyms(string intType, string values, string sqlInsertValues, long?[] expectedValues)
    {
        _ = values; // used for test display name only

        // Given Snowflake client is logged in
        using var connection = Fixture.Factory.Create(Output);
        connection.Open();

        // And Table with <type> column exists with values <insert_values>
        var tableName = $"UD_INT_TBL_{Guid.NewGuid():N}".Substring(0, 30);
        using var createCmd = connection.CreateCommand();
        createCmd.CommandText = $"CREATE TEMPORARY TABLE {tableName} (col {intType})";
        createCmd.ExecuteNonQuery();

        using var insertCmd = connection.CreateCommand();
        insertCmd.CommandText = $"INSERT INTO {tableName} (col) VALUES {sqlInsertValues}";
        insertCmd.ExecuteNonQuery();

        // When Query "SELECT * FROM <table> ORDER BY col" is executed
        using var selectCmd = connection.CreateCommand();
        selectCmd.CommandText = $"SELECT * FROM {tableName} ORDER BY col";
        using var reader = selectCmd.ExecuteReader();

        // Then Result should contain integers <expected_values>
        var rowIndex = 0;
        while (reader.Read())
        {
            if (expectedValues[rowIndex] is null)
            {
                Assert.True(reader.IsDBNull(0), $"Expected NULL at row {rowIndex}");
            }
            else
            {
                Assert.Equal(expectedValues[rowIndex]!.Value, reader.GetInt64(0));
            }
            rowIndex++;
        }
        Assert.Equal(expectedValues.Length, rowIndex);
    }

    public static IEnumerable<object[]> TableValuesData()
    {
        foreach (var intType in IntTypeSynonyms)
        {
            yield return [
                intType,
                "positive",
                "(0), (1), (127), (255), (32767), (65535), (2147483647), (4294967295), (9223372036854775807)",
                new long?[] { 0L, 1L, 127L, 255L, 32767L, 65535L, 2147483647L, 4294967295L, 9223372036854775807L }
            ];
            yield return [
                intType,
                "negative",
                "(-1), (-128), (-32768), (-2147483648), (-9223372036854775808)",
                new long?[] { -9223372036854775808L, -2147483648L, -32768L, -128L, -1L }
            ];
            yield return [
                intType,
                "null",
                "(0), (NULL), (42)",
                new long?[] { 0L, 42L, null }
            ];
        }
    }

    // Scenario: should handle server-side Arrow memory optimization for int columns on multiple chunks
    [SnowflakeFact]
    public void ShouldHandleServerSideArrowMemoryOptimizationForIntColumnsOnMultipleChunks()
    {
        const int rowCount = 50000;

        // Given Snowflake client is logged in
        using var connection = Fixture.Factory.Create(Output);
        connection.Open();

        // And Table with four INT columns exists
        var tableName = $"UD_INT_ARROW_{Guid.NewGuid():N}".Substring(0, 30);
        using var createCmd = connection.CreateCommand();
        createCmd.CommandText = $"CREATE TEMPORARY TABLE {tableName} (col_int8 INT, col_int16 INT, col_int32 INT, col_int64 INT)";
        createCmd.ExecuteNonQuery();

        // And Each column contains values of different magnitudes (50000 rows to span multiple Arrow chunks)
        using var insertCmd = connection.CreateCommand();
        insertCmd.CommandText = $"INSERT INTO {tableName} SELECT 100, 30000, 2000000000, 9000000000000000000 FROM TABLE(GENERATOR(ROWCOUNT => {rowCount}))";
        insertCmd.ExecuteNonQuery();

        // When Query "SELECT * FROM <table>" is executed
        using var selectCmd = connection.CreateCommand();
        selectCmd.CommandText = $"SELECT * FROM {tableName}";
        using var reader = selectCmd.ExecuteReader();

        // Then Result should contain 50000 rows with all values equal to expected data
        var count = 0;
        while (reader.Read())
        {
            Assert.Equal(100L, reader.GetInt64(0));
            Assert.Equal(30000L, reader.GetInt64(1));
            Assert.Equal(2000000000L, reader.GetInt64(2));
            Assert.Equal(9000000000000000000L, reader.GetInt64(3));
            count++;
        }
        Assert.Equal(rowCount, count);
    }

    // Scenario: should insert integer using parameter binding for int and synonyms
    [SnowflakeTheory]
    [InlineData("INT")]
    [InlineData("INTEGER")]
    [InlineData("BIGINT")]
    [InlineData("SMALLINT")]
    [InlineData("TINYINT")]
    [InlineData("BYTEINT")]
    public void ShouldInsertIntegerUsingParameterBindingForIntAndSynonyms(string intType)
    {
        Skip.FutureMilestone();
        long[] testValues = [0, -2147483648, 2147483647, 9223372036854775807];

        // Given Snowflake client is logged in
        using var connection = Fixture.Factory.Create(Output);
        connection.Open();

        // And Table with <type> column exists
        var tableName = $"UD_INT_BIND_{Guid.NewGuid():N}".Substring(0, 30);
        using var createCmd = connection.CreateCommand();
        createCmd.CommandText = $"CREATE TEMPORARY TABLE {tableName} (col {intType})";
        createCmd.ExecuteNonQuery();

        // When Integer values [0, -2147483648, 2147483647, 9223372036854775807] are inserted using binding
        foreach (var value in testValues)
        {
            using var insertCmd = connection.CreateCommand();
            insertCmd.CommandText = $"INSERT INTO {tableName} (col) VALUES (?)";
            var param = insertCmd.CreateParameter();
            param.ParameterName = "1";
            param.DbType = DbType.Int64;
            param.Value = value;
            insertCmd.Parameters.Add(param);
            insertCmd.ExecuteNonQuery();
        }

        // And Query "SELECT * FROM <table>" is executed
        using var selectCmd = connection.CreateCommand();
        selectCmd.CommandText = $"SELECT col FROM {tableName} ORDER BY col";
        using var reader = selectCmd.ExecuteReader();

        // Then Result should contain integers [0, -2147483648, 2147483647, 9223372036854775807]
        var sorted = testValues.OrderBy(x => x).ToArray();
        for (var i = 0; i < sorted.Length; i++)
        {
            Assert.True(reader.Read(), $"Expected row {i}");
            Assert.Equal(sorted[i], reader.GetInt64(0));
        }
        Assert.False(reader.Read(), $"Expected exactly {sorted.Length} rows");
    }

    // Scenario: should insert and select integers from table using batch parameter binding for int and synonyms
    [SnowflakeTheory]
    [InlineData("INT")]
    [InlineData("INTEGER")]
    [InlineData("BIGINT")]
    [InlineData("SMALLINT")]
    [InlineData("TINYINT")]
    [InlineData("BYTEINT")]
    public void ShouldInsertAndSelectIntegersFromTableUsingBatchParameterBindingForIntAndSynonyms(string intType)
    {
        Skip.FutureMilestone();
        long[] testValues = [0, 42, -2147483648, 2147483647, 9223372036854775807];

        // Given Snowflake client is logged in
        using var connection = Fixture.Factory.Create(Output);
        connection.Open();

        // And Table with <type> column exists
        var tableName = $"UD_INT_BATCH_{Guid.NewGuid():N}".Substring(0, 30);
        using var createCmd = connection.CreateCommand();
        createCmd.CommandText = $"CREATE TEMPORARY TABLE {tableName} (col {intType})";
        createCmd.ExecuteNonQuery();

        // When Integer values [0, 42, -2147483648, 2147483647, 9223372036854775807] are inserted using binding
        foreach (var value in testValues)
        {
            using var insertCmd = connection.CreateCommand();
            insertCmd.CommandText = $"INSERT INTO {tableName} (col) VALUES (?)";
            var param = insertCmd.CreateParameter();
            param.ParameterName = "1";
            param.DbType = DbType.Int64;
            param.Value = value;
            insertCmd.Parameters.Add(param);
            insertCmd.ExecuteNonQuery();
        }

        // And Query "SELECT * FROM <table>" is executed
        using var selectCmd = connection.CreateCommand();
        selectCmd.CommandText = $"SELECT col FROM {tableName} ORDER BY col";
        using var reader = selectCmd.ExecuteReader();

        // Then Result should contain integers [0, 42, -2147483648, 2147483647, 9223372036854775807]
        var sorted = testValues.OrderBy(x => x).ToArray();
        for (var i = 0; i < sorted.Length; i++)
        {
            Assert.True(reader.Read(), $"Expected row {i}");
            Assert.Equal(sorted[i], reader.GetInt64(0));
        }
        Assert.False(reader.Read(), $"Expected exactly {sorted.Length} rows");
    }

    // Scenario: should handle large integer values as string for int and synonyms
    [SnowflakeTheory]
    [InlineData("INT")]
    [InlineData("INTEGER")]
    [InlineData("BIGINT")]
    [InlineData("SMALLINT")]
    [InlineData("TINYINT")]
    [InlineData("BYTEINT")]
    public void ShouldHandleLargeIntegerValuesAsStringForIntAndSynonyms(string intType)
    {
        Skip.FutureMilestone();

        // Given Snowflake client is logged in
        using var connection = Fixture.Factory.Create(Output);
        connection.Open();

        // When Query "SELECT -99999999999999999999999999999999999999::<type>, 99999999999999999999999999999999999999::<type>" is executed
        using var cmd = connection.CreateCommand();
        cmd.CommandText = $"SELECT -99999999999999999999999999999999999999::{intType}, 99999999999999999999999999999999999999::{intType}";
        using var reader = cmd.ExecuteReader();

        // Then Result should contain string values ["-99999999999999999999999999999999999999", "99999999999999999999999999999999999999"]
        Assert.True(reader.Read(), "Expected one row");
        Assert.Equal("-99999999999999999999999999999999999999", reader.GetString(0));
        Assert.Equal("99999999999999999999999999999999999999", reader.GetString(1));
        Assert.False(reader.Read(), "Expected exactly one row");
    }
}

File: nodejs/tests/e2e/query/data-types/int.test.ts
import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import type { Connection } from '../../../types/sdk-types.js';
import { createLiveConnection, createTemporaryTable } from '../../utils/fixtures.js';
import {
  destroyConnectionAsync,
  executeAsync,
  getStatementColumn,
  NOT_IMPLEMENTED_IN_NEW_DRIVER,
} from '../../utils/index.js';
import {
  createLiveNullPreservingConnection,
  expectBigIntValue,
  expectBigIntValues,
} from '../utils.js';

describe('INT data type', () => {
  describe('tests/definitions/shared/types/int.feature', () => {
    let connection: Connection;

    beforeAll(async () => {
      // Default driver behavior is losing precision for anything higher than Number.MAX_SAFE_INTEGER
      // So we use jsTreatIntegerAsBigInt and assert every value as BigInt
      connection = await createLiveConnection({ jsTreatIntegerAsBigInt: true }, false);
    });

    afterAll(async () => {
      await destroyConnectionAsync(connection);
    });

    it.each(['INT', 'INTEGER', 'BIGINT', 'SMALLINT', 'TINYINT', 'BYTEINT'])(
      'should cast integer values to appropriate type for int and synonyms (%s)',
      async (type) => {
        // Given Snowflake client is logged in
        void connection;

        // When Query "SELECT 0::<type>, 1000000::<type>, 9223372036854775807::<type>" is executed
        const { statement, rows } = await executeAsync(
          connection,
          `SELECT 0::${type}, 1000000::${type}, 9223372036854775807::${type}`,
        );

        // Then All values should be returned as appropriate type with no precision loss
        for (const index of [0, 1, 2]) {
          const column = getStatementColumn(statement, index);
          expect(column.getType()).toBe('fixed');
          expect(column.isNumber()).toBe(true);
          expect(column.getScale()).toBe(0);
        }
        expectBigIntValues(Object.values(rows[0]), ['0', '1000000', '9223372036854775807']);
      },
    );

    it('should select integer <values> for int and synonyms', async () => {
      // Given Snowflake client is logged in
      void connection;

      const literalCases: [query: string, expected: string[]][] = [
        ['0::INT', ['0']],
        ['-128::INT, 127::INT, 255::INT', ['-128', '127', '255']],
        ['-32768::INT, 32767::INT, 65535::INT', ['-32768', '32767', '65535']],
        [
          '-2147483648::INT, 2147483647::INT, 4294967295::INT',
          ['-2147483648', '2147483647', '4294967295'],
        ],
        [
          '-9223372036854775808::INT, 9223372036854775807::INT',
          ['-9223372036854775808', '9223372036854775807'],
        ],
      ];

      for (const [query, expected] of literalCases) {
        // When Query "SELECT <query_values>" is executed
        const { rows } = await executeAsync(connection, `SELECT ${query}`);

        // Then Result should contain integers <expected_values>
        expectBigIntValues(Object.values(rows[0]), expected);
      }
    });

    it('should handle large integer values for int and synonyms', async () => {
      // Given Snowflake client is logged in
      void connection;

      // When Query "SELECT -99999999999999999999999999999999999999::<type>, 99999999999999999999999999999999999999::<type>" is executed
      const { rows } = await executeAsync(
        connection,
        `SELECT
           -99999999999999999999999999999999999999::INT,
           99999999999999999999999999999999999999::INT`,
      );

      // Then Result should contain integers [-99999999999999999999999999999999999999, 99999999999999999999999999999999999999]
      expectBigIntValues(Object.values(rows[0]), [
        '-99999999999999999999999999999999999999',
        '99999999999999999999999999999999999999',
      ]);
    });

    it('should handle NULL values for int and synonyms', async () => {
      // Given Snowflake client is logged in
      void connection;

      // When Query "SELECT NULL::<type>, 42::<type>, NULL::<type>" is executed
      const { rows } = await executeAsync(
        connection,
        `SELECT NULL::INT, 42::INT, NULL::INT AS NULL_COL2`,
      );

      // Then Result should contain [NULL, 42, NULL]
      const [c1, c2, c3] = Object.values(rows[0]);
      expect(c1).toBeNull();
      expectBigIntValue(c2, '42');
      expect(c3).toBeNull();
    });

    it('should select <values> from table for int and synonyms', async () => {
      // Given Snowflake client is logged in
      void connection;

      const tableCases: [insert: string, expected: (string | null)[]][] = [
        [
          '(0), (1), (127), (255), (32767), (65535), (2147483647), (4294967295), (9223372036854775807)',
          [
            '0',
            '1',
            '127',
            '255',
            '32767',
            '65535',
            '2147483647',
            '4294967295',
            '9223372036854775807',
          ],
        ],
        [
          '(-1), (-128), (-32768), (-2147483648), (-9223372036854775808)',
          ['-9223372036854775808', '-2147483648', '-32768', '-128', '-1'],
        ],
        ['(0), (NULL), (42)', ['0', '42', null]],
      ];

      for (const [insert, expected] of tableCases) {
        // And Table with <type> column exists with values <insert_values>
        const tableName = await createTemporaryTable(connection, 'COL INT');
        await executeAsync(connection, `INSERT INTO ${tableName} (COL) VALUES ${insert}`);

        // When Query "SELECT * FROM <table> ORDER BY col" is executed
        const { rows } = await executeAsync(
          connection,
          `SELECT COL FROM ${tableName} ORDER BY COL NULLS LAST`,
        );

        // Then Result should contain integers <expected_values>
        expect(rows).toHaveLength(expected.length);
        expected.forEach((rowValue, index) => {
          if (rowValue === null) {
            expect(rows[index].COL).toBeNull();
          } else {
            expectBigIntValue(rows[index].COL, rowValue);
          }
        });
      }
    });

    it('should select large integer values from table for int and synonyms', async () => {
      // Given Snowflake client is logged in
      void connection;

      // And Table with <type> column exists with values [-99999999999999999999999999999999999999, 99999999999999999999999999999999999999]
      const tableName = await createTemporaryTable(connection, 'COL INT');
      await executeAsync(
        connection,
        `INSERT INTO ${tableName} (COL) VALUES
           (-99999999999999999999999999999999999999),
           (99999999999999999999999999999999999999)`,
      );

      // When Query "SELECT * FROM <table> ORDER BY col" is executed
      const { rows } = await executeAsync(connection, `SELECT COL FROM ${tableName} ORDER BY COL`);

      // Then Result should contain integers [-99999999999999999999999999999999999999, 99999999999999999999999999999999999999]
      expectBigIntValues(
        rows.map((row) => row.COL),
        ['-99999999999999999999999999999999999999', '99999999999999999999999999999999999999'],
      );
    });

    it('should insert integer using parameter binding for int and synonyms', async () => {
      // Given Snowflake client is logged in
      void connection;

      // And Table with <type> column exists
      const tableName = await createTemporaryTable(connection, 'COL INT');

      // When Integer values [0, -2147483648, 2147483647, 9223372036854775807] are inserted using binding
      await executeAsync(connection, `INSERT INTO ${tableName} (COL) VALUES (?)`, {
        binds: [[0], [-2147483648], [2147483647], ['9223372036854775807']],
      });

      // And Query "SELECT * FROM <table>" is executed
      const { rows } = await executeAsync(connection, `SELECT COL FROM ${tableName} ORDER BY COL`);

      // Then Result should contain integers [0, -2147483648, 2147483647, 9223372036854775807]
      expectBigIntValues(
        rows.map((row) => row.COL),
        ['-2147483648', '0', '2147483647', '9223372036854775807'],
      );
    });

    describe.skipIf(NOT_IMPLEMENTED_IN_NEW_DRIVER)('multiple chunks', () => {
      const ROW_COUNT = 50_000;

      it('should download large result set with multiple chunks for int and synonyms', async () => {
        // Given Snowflake client is logged in
        void connection;

        // When Query "SELECT seq8()::<type> as id FROM TABLE(GENERATOR(ROWCOUNT => 50000)) v ORDER BY id" is executed
        const { rows } = await executeAsync(
          connection,
          `SELECT seq8()::INT AS ID FROM TABLE(GENERATOR(ROWCOUNT => ${ROW_COUNT})) v ORDER BY ID`,
        );

        // Then Result should contain 50000 sequentially numbered rows from 0 to 49999
        expect(rows).toHaveLength(ROW_COUNT);
        expectBigIntValues([rows[0].ID, rows[ROW_COUNT - 1].ID], ['0', String(ROW_COUNT - 1)]);
      });

      it('should handle server-side Arrow memory optimization for int columns on multiple chunks', async () => {
        // Given Snowflake client is logged in
        void connection;

        // And Table with four INT columns exists
        const tableName = await createTemporaryTable(
          connection,
          'COL_INT8 INT, COL_INT16 INT, COL_INT32 INT, COL_INT64 INT',
        );

        // And Each column contains values of different magnitudes (50000 rows to span multiple Arrow chunks)
        await executeAsync(
          connection,
          `INSERT INTO ${tableName} (COL_INT8, COL_INT16, COL_INT32, COL_INT64)
             SELECT
               (seq8() % 256) - 128,
               (seq8() % 65536) - 32768,
               (seq8() % 4294967296) - 2147483648,
               seq8() * 100000000000
             FROM TABLE(GENERATOR(ROWCOUNT => ${ROW_COUNT})) v`,
        );

        // When Query "SELECT * FROM <table>" is executed
        const { rows } = await executeAsync(
          connection,
          `SELECT COL_INT8, COL_INT16, COL_INT32, COL_INT64 FROM ${tableName}
             ORDER BY COL_INT64`,
        );

        // Then Result should contain 50000 rows with all values equal to expected data
        expect(rows).toHaveLength(ROW_COUNT);
        expectBigIntValues(Object.values(rows[0]), ['-128', '-32768', '-2147483648', '0']);
      });
    });
  });

  describe('fetchAsString', () => {
    it.each([true, false])(
      'should return integers beyond MAX_SAFE_INTEGER as exact digit strings without precision loss (jsTreatIntegerAsBigInt=%s)',
      async (jsTreatIntegerAsBigInt) => {
        const connection = await createLiveConnection({
          jsTreatIntegerAsBigInt,
        });

        // MAX_SAFE_INTEGER + 4 rounds to a different Number, so the string form proves precision is kept.
        const beyondSafeInteger = BigInt(Number.MAX_SAFE_INTEGER) + 4n;
        expect(String(Number(beyondSafeInteger))).not.toBe(String(beyondSafeInteger));

        const { rows } = await executeAsync(
          connection,
          `SELECT 123::INT AS C1, ${beyondSafeInteger}::INT, 1e1::INT`,
          { fetchAsString: ['Number'] },
        );
        const [small, large, exponential] = Object.values(rows[0]);
        expect([small, large, exponential]).toEqual(['123', String(beyondSafeInteger), '10']);
      },
    );

    it("should render a NULL INT cell as the string 'NULL'", async () => {
      const connection = await createLiveConnection({});
      const { rows } = await executeAsync(connection, 'SELECT NULL::INT', {
        fetchAsString: ['Number'],
      });
      expect(Object.values(rows[0])).toEqual(['NULL']);
    });

    it('should render a NULL INT cell as null when representNullAsStringNull is disabled', async () => {
      const nullPreservingConnection = await createLiveNullPreservingConnection();
      const { rows } = await executeAsync(nullPreservingConnection, 'SELECT NULL::INT', {
        fetchAsString: ['Number'],
      });
      expect(Object.values(rows[0])).toEqual([null]);
    });
  });
});
